src/coreclr/tools/Common/TypeSystem/Common/MetadataVirtualMethodAlgorithm.cs

```diff
@@ -1003,11 +1003,14 @@ private static MethodDesc TryResolveVirtualStaticMethodOnThisType(MetadataType c
                 if (methodImpl.Decl == interfaceMethodDefinition)
                 {
                     MethodDesc resolvedMethodImpl = methodImpl.Body;
-                    if (interfaceMethod != interfaceMethodDefinition)
+                    if (resolvedMethodImpl.OwningType == constrainedType.GetTypeDefinition())
```

Comment:
Just to double check - does this match what the VM will do? This code also runs on Native AOT where we cannot just let things fall back to JIT. If this _can_ resolve, it _must_ resolve. I want to make sure that in order to fix a test with invalid IL with crossgen2, we're not making valid IL fail to compile with Native AOT.

Per ECMA-335 "The method indexed by MethodBody shall be a member of Class or some base class of Class (MethodImpls do not allow compilers to ‘hook’ arbitrary method bodies)". We are rejecting the "some base class of Class" here.

Reply:
Thanks for pointing that out, I have updated the algorithm to properly account for base types, this also fixed the six failures in typesystem tests I saw in my yesterday lab runs.